build-system-abstraction.rst

```diff
@@ -0,0 +1,472 @@
+:PEP: XX
+:Title: Build system abstraction for pip/conda etc
+:Version: $Revision$
+:Last-Modified: $Date$
+:Author: Robert Collins <rbtcollins@hp.com>,
+         Nathaniel Smith <njs@pobox.com>
+:BDFL-Delegate: Donald Stufft <donald@stufft.io>
+:Discussions-To: distutils-sig <distutils-sig@python.org>
+:Status: Draft
+:Type: Standards Track
+:Content-Type: text/x-rst
+:Created: 26-Oct-2015
+:Post-History: XX
+:Requires: The new dependency-specification PEP.
+
+
+Abstract
+========
+
+This PEP specifies a programmatic interface for pip [#pip]_ and other
+distribution or installation tools to use when working with Python
+source trees (both the developer tree - e.g. the git tree - and source
+distributions).
+
+The programmatic interface allows decoupling of pip from its current
+hard dependency on setuptools [#setuptools]_ able for two
+key reasons:
+
+1. It enables new build systems that may be much easier to use without
+   requiring them to even appear to be setuptools.
+
+2. It facilitates setuptools itself changing its user interface without
+   breaking pip, giving looser coupling.
+
+The interface needed to permit pip to install build systems also enables pip to
+install build time requirements for packages which is an important step in
+getting pip to full feature parity with the installation components of
+easy-install.
+
+As PEP-426 [#pep426]_ is draft, we cannot utilise the metadata format it
+defined. However PEP-427 wheels are in wide use and fairly well specified, so
+we have adopted the METADATA format from that for specifying distribution
+dependencies and general project metadata. PEP-0508 [#pep508] provides a self
+contained language for describing a dependency, which we encapsulate in a thin
+JSON schema to describe bootstrap dependencies.
+
+Motivation
+==========
+
+There is significant pent-up frustration in the Python packaging ecosystem
+around the current lock-in between build system and pip. Breaking that lock-in
+is better for pip, for setuptools, and for other build systems like flit
+[#flit]_.
+
+Specification
+=============
+
+Overview
+--------
+
+Build tools will be located by reading a file ``pypa.json`` from the root
+directory of the source tree. That file describes how to get the build tool
+and the name of the command to run to invoke the tool.
+
+All tools will be expected to conform to a single command line interface
+modelled on pip's existing use of the setuptools setup.py interface.
+
+pypa.json
+---------
+
+The file ``pypa.json`` acts as neutral configuration file for pip and other
+tools that want to build source trees to consult for configuration. The
+absence of a ``pypa.json`` file in a Python source tree implies a setuptools
+or setuptools compatible build system.
+
+The JSON has the following schema. Extra keys are ignored, which permits the
+use of ``pypa.json`` as a configuration file for other related tools. If doing
+that the chosen keys must be namespaced - e.g. ``flit`` with keys under that
```

Comment:
If the format is JSON, I *don't* think we should do this. I was only suggesting it if we decided to switch to something like TOML/YAML. I'm very happy to leave it as JSON, but in this case, I wouldn't allow build tools to put extra data in this file. JSON is good precisely because the file is small and simple.

Also, I don't think we should allow build tools to add even one arbitrary key at the top level, because then there's no way to have an 'unknown key' error. If we do allow this, I think we should pick one top-level key e.g. `buildtool` and require that build tools put their data under that. But again, if it's JSON, let's not bother - I'd rather stick with a separate file.